src/audio/google/google_rtc_audio_processing.c

```diff
@@ -639,6 +733,180 @@ static int google_rtc_audio_processing_reset(struct processing_module *mod)
 	return 0;
 }
 
+#if CONFIG_IPC_MAJOR_4
+static inline int16_t convert_google_aec_format_to_int16(BUF_TYPE data)
+{
+#if CONFIG_COMP_GOOGLE_RTC_USE_32_BIT_FLOAT_API
+#if XCHAL_HAVE_HIFI3
+	const xtfloat ratio = 2 << 14;
+	xtfloat x0 = data;
+	xtfloat x1;
+	int16_t x;
+
+	x1 = XT_MUL_S(x0, ratio);
+	x = XT_TRUNC_S(x1, 0);
+
+	return x;
+#else /* XCHAL_HAVE_HIFI3 */
+	return Q_CONVERT_FLOAT(data, 15);
+#endif /* XCHAL_HAVE_HIFI3 */
+#else /* CONFIG_COMP_GOOGLE_RTC_USE_32_BIT_FLOAT_API */
+	return data;
+#endif
+}
+
+static inline BUF_TYPE convert_int16_to_google_aec_format(int16_t data)
+{
+#if CONFIG_COMP_GOOGLE_RTC_USE_32_BIT_FLOAT_API
+#if XCHAL_HAVE_HIFI3
+	const xtfloat ratio = 2 << 14;
+	xtfloat x0 = data;
+	float x;
+
+	x = XT_DIV_S(x0, ratio);
+
+	return x;
+#else /* XCHAL_HAVE_HIFI3 */
+	return Q_CONVERT_QTOF(data, 15);
+#endif /* XCHAL_HAVE_HIFI3 */
+#else /* CONFIG_COMP_GOOGLE_RTC_USE_32_BIT_FLOAT_API */
+	return data;
+#endif
+}
+
+/* todo CONFIG_FORMAT_S32LE */
+static int google_rtc_audio_processing_process(struct processing_module *mod,
+					       struct sof_source **sources, int num_of_sources,
+					       struct sof_sink **sinks, int num_of_sinks)
+{
+	int ret;
+	int16_t const *src;
+	int8_t const *src_buf_start;
+	int8_t const *src_buf_end;
+	size_t src_buf_size;
+
+	int16_t const *ref;
+	int8_t const *ref_buf_start;
+	int8_t const *ref_buf_end;
+	size_t ref_buf_size;
+
+	int16_t *dst;
+	int8_t *dst_buf_start;
+	int8_t *dst_buf_end;
+	size_t dst_buf_size;
+
+	size_t num_of_bytes_to_process;
+	size_t channel;
+
+	struct sof_source *ref_stream, *src_stream;
+	struct sof_sink *dst_stream;
+
+	struct google_rtc_audio_processing_comp_data *cd = module_get_private_data(mod);
+
+	if (cd->reconfigure) {
+		ret = google_rtc_audio_processing_reconfigure(mod);
+		if (ret)
+			return ret;
+	}
+
+	src_stream = sources[cd->raw_microphone_source];
+	ref_stream = sources[cd->aec_reference_source];
+	dst_stream = sinks[0];
+
+	num_of_bytes_to_process = cd->num_frames * source_get_frame_bytes(ref_stream);
+	ret = source_get_data(ref_stream, num_of_bytes_to_process, (const void **)&ref,
+			      (const void **)&ref_buf_start, &ref_buf_size);
```

Comment:
DP should NOT process more than its period

why
 - DP start in async to LL, so it may happen that between DP becomes ready (10ms in each of buffers),  there's another LL tick and one of the pipelines will provide additional data portions. 

TICK
 - Core 0 LL finishes processing, AEC has 10ms in both buffers => ready to running
 TICK 
 - core 0 starts another LL cycle, core1 starts DP (and is late a bit- i.e. was processing IPC, does not matter why)
 - **core0 finishes some processing before DP starts**, there's additional 1ms data in one of inputs (or even on both)

but "additional 1ms" belong to a next cycle. Process it in next cycle

Reply:
Thank you for the clarification.